src/audio/drc/drc_algorithm_hifi3.c

```diff
@@ -0,0 +1,469 @@
+	if (nbyte == 2) { /* 2 bytes per sample */
+		for (ch = 0; ch < nch; ch++) {
+			abs_input_array_p = abs_input_array;
+			sample16_p = (int16_t *)state->pre_delay_buffers[ch] + div_start;
+			for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+				sample = (int32_t)*sample16_p << 16;
+				*abs_input_array_p = MAX(*abs_input_array_p, ABS(sample));
+				abs_input_array_p++;
+				sample16_p++;
+			}
+		}
+	} else { /* 4 bytes per sample */
+		for (ch = 0; ch < nch; ch++) {
+			abs_input_array_p = abs_input_array;
+			sample32_p = (int32_t *)state->pre_delay_buffers[ch] + div_start;
+			for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+				sample = *sample32_p;
+				*abs_input_array_p = MAX(*abs_input_array_p, ABS(sample));
+				abs_input_array_p++;
+				sample32_p++;
+			}
+		}
+	}
+
+	for (i = 0; i < DRC_DIVISION_FRAMES; i++) {
+		/* Compute compression amount from un-delayed signal */
+
+		/* Calculate shaped power on undelayed input.  Put through
+		 * shaping curve. This is linear up to the threshold, then
+		 * enters a "knee" portion followed by the "ratio" portion. The
+		 * transition from the threshold to the knee is smooth (1st
+		 * derivative matched). The transition from the knee to the
+		 * ratio portion is smooth (1st derivative matched).
+		 */
+		gain = volume_gain(p, abs_input_array[i]); /* Q2.30 */
+		gain_diff = AE_SUB32(gain, detector_average); /* Q2.30 */
+		is_release = ((int32_t)gain_diff > 0);
+		if (is_release) {
+			if ((int32_t)gain > NEG_TWO_DB_Q30) {
+				tmp = mult(gain_diff, p->sat_release_rate_at_neg_two_db,
+					   30, 30, 30);
+			} else {
+				gain = AE_SRAI32R(gain, 4); /* Q2.30 -> Q6.26 */
+				db_per_frame = mult(drc_lin2db_fixed(gain),
+						    p->sat_release_frames_inv_neg, 21, 30, 24);
+				sat_release_rate = AE_SUB32(db2lin_fixed(db_per_frame), ONE_Q20);
+				tmp = mult(gain_diff, sat_release_rate, 30, 20, 30);
+			}
+			detector_average = AE_ADD32(detector_average, tmp);
+		} else {
+			detector_average = gain;
+		}
+
+		detector_average = AE_MIN32(detector_average, ONE_Q30);
+	}
+
+	state->detector_average = detector_average;
+}
+
+/* Updates the envelope_rate used for the next division */
+void drc_update_envelope(struct drc_state *state, const struct sof_drc_params *p)
+{
+	/* Calculate desired gain */
+
+	/* Pre-warp so we get desired_gain after sin() warp below. */
+	int32_t scaled_desired_gain = drc_asin_fixed(state->detector_average); /* Q2.30 */
+
+	/* Deal with envelopes */
+
+	/* envelope_rate is the rate we slew from current compressor level to
+	 * the desired level.  The exact rate depends on if we're attacking or
+	 * releasing and by how much.
+	 */
+	ae_f32 envelope_rate;
+
+	int is_releasing = scaled_desired_gain > state->compressor_gain;
+
+	/* compression_diff_db is the difference between current compression
+	 * level and the desired level.
+	 */
+	int is_bad_db = (state->compressor_gain == 0 || scaled_desired_gain == 0);
+	ae_f32 compression_diff_db;
```

Comment:
Single declarations to be able quickly see/count ae registers usage. Also order, there's a normal int32_t in the middle.